File: 1334-find-the-city-with-the-smallest-number-of-neighbors-at-a-threshold-distance/1334-find-the-city-with-the-smallest-number-of-neighbors-at-a-threshold-distance.py
class Solution:
    def findTheCity(self, n: int, edges: List[List[int]], distanceThreshold: int) -> int:
        graph = {i:dict() for i in range(n)}
        for u,v,w in edges:
            graph[u][v] = w
            graph[v][u] = w
        
        neighbors = [0]*n
        
        for k in range(n):
            dist = [float('inf')]*n
            dist[k] = 0
            visited = [False]*n
            
            queue = [(0, k)]
            heapify(queue)
            
            count = -1

            while len(queue):
                minVal, minNode = heappop(queue)
                if minVal > distanceThreshold: break
                if visited[minNode]: continue
                visited[minNode] = True
                count += 1
                for node in graph[minNode]:
                    if not visited[node] and dist[minNode] + graph[minNode][node] < dist[node]:
                        dist[node] = dist[minNode] + graph[minNode][node]
                        heappush(queue, (dist[node], node))

            neighbors[k] = count
            
        curMin = neighbors[0]
        ans = 0
        for i in range(n):
            if neighbors[i] <= curMin:
                ans = i
                curMin = neighbors[i]

        return ans

File: 1334-find-the-city-with-the-smallest-number-of-neighbors-at-a-threshold-distance/1334-find-the-city-with-the-smallest-number-of-neighbors-at-a-threshold-distance.cpp
class Solution {
public:
    int findTheCity(int n, vector<vector<int>>& edges, int distanceThreshold) {
        vector<vector<int>> cost(n+1,vector<int>(n+1,INT_MAX));
        for(auto x:edges)
        {
            cost[x[0]][x[1]]=x[2];
            cost[x[1]][x[0]]=x[2];
        }
        for(int k=0;k<n;k++)
        {
            for(int i=0;i<n;i++)
            {
                if(cost[i][k]!=INT_MAX)
                for(int j=0;j<n;j++)
                {
                    if(i!=j && cost[k][j]!=INT_MAX)
                        cost[i][j]=min(cost[i][j],cost[i][k]+cost[k][j]);
                }
            }
        }
        int res=0;
        int m=n;
        for(int i=0;i<n;i++)
        {
            int c=0;
            for(int j=0;j<n;j++)
            {
                if(cost[i][j]<=distanceThreshold && cost[i][j]!=INT_MAX)
                    c++;
            }
            if(c<=m)
            {
                m=c;
                res=i;
            }
        }
        return res;
    }
};